lib/resty/core/regex.lua

```diff
@@ -478,6 +479,87 @@ function ngx.re.find(subj, regex, opts, ctx, nth)
     return re_match_helper(subj, regex, opts, ctx, false, nil, nth)
 end
 
+local function destroy_re_gmatch_iterator(iterator)
+    if not iterator._compile_once then
+        destroy_compiled_regex(iterator._compiled)
+    end
+    iterator._compiled = nil
+    iterator._pos = nil
+    iterator._subj = nil
+end
+
+local function iterate_re_gmatch(self)
+    local compiled = self._compiled
+    local subj = self._subj
+    local subj_len = self._subj_len
+    local flags = self._flags
+    local pos = self._pos
+
+    if not pos then
+        -- The iterator is exhausted.
+        return nil
+    end
+
+    local rc = C.ngx_http_lua_ffi_exec_regex(compiled, flags, subj,
+                                             subj_len, pos)
+
+    if rc == PCRE_ERROR_NOMATCH then
+        destroy_re_gmatch_iterator(self)
+        return nil
+    end
+
+    if rc < 0 then
+        destroy_re_gmatch_iterator(self)
+        return nil, "pcre_exec() failed: " .. rc
+    end
+
+    if rc == 0 then
+        if band(flags, FLAG_DFA) == 0 then
+            destroy_re_gmatch_iterator(self)
+            return nil, "capture size too small"
+        end
+
+        rc = 1
+    end
+
+    local cp_pos = tonumber(compiled.captures[1])
+    if cp_pos == compiled.captures[0] then
+        cp_pos = cp_pos + 1
+        if cp_pos > subj_len then
+            local res = collect_captures(compiled, rc, subj, flags)
+            destroy_re_gmatch_iterator(self)
+            return res
+        end
+    end
+    self._pos = cp_pos
+    return collect_captures(compiled, rc, subj, flags)
+end
+
+
+do
```

Comment:
The 2 above local functions could be declared inside this `do ... end` block. What do you think?

Reply:
Agree.